cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,209 @@
+#include <cugraph/graph.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/thrust_tuple_utils.hpp>
+#include <cugraph/vertex_partition_device_view.cuh>
+
+#include <sampling/detail/graph_functions.hpp>
+
+#include <raft/core/handle.hpp>
+
+#include <thrust/optional.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+#include <rmm/device_uvector.hpp>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename label_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<label_t>>,
+           std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                                    std::optional<rmm::device_uvector<label_t>>>>>
+prepare_next_frontier(
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> starting_vertices,
+  std::optional<raft::device_span<label_t const>> starting_vertex_labels,
+  raft::device_span<vertex_t const> new_frontier_vertices,
+  std::optional<raft::device_span<label_t const>> new_frontier_vertex_labels,
+  std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                           std::optional<rmm::device_uvector<label_t>>>>&& vertex_used_as_source,
+  vertex_partition_view_t<vertex_t, multi_gpu> vertex_partition,
+  std::vector<vertex_t> const& vertex_partition_range_lasts,
+  bool carry_over_sources,
+  bool dedupe_sources,
+  bool do_expensive_check)
+{
+  vertex_partition_device_view_t<vertex_t, multi_gpu> d_vertex_partition(vertex_partition);
+
+  if (vertex_used_as_source) {
+    auto& [verts, labels] = *vertex_used_as_source;
+
+    size_t current_size = verts.size();
+    verts.resize(current_size + starting_vertices.size(), handle.get_stream());
+    raft::copy(verts.begin() + current_size,
+               starting_vertices.data(),
+               starting_vertices.size(),
+               handle.get_stream());
+
+    if (starting_vertex_labels) {
+      labels->resize(current_size + starting_vertex_labels->size(), handle.get_stream());
+
+      raft::copy(labels->begin() + current_size,
+                 starting_vertex_labels->data(),
+                 starting_vertex_labels->size(),
+                 handle.get_stream());
+
+      auto begin_iter = thrust::make_zip_iterator(verts.begin(), labels->begin());
+
+      thrust::sort(handle.get_thrust_policy(), begin_iter, begin_iter + verts.size());
+
+      auto new_end =
+        thrust::unique(handle.get_thrust_policy(), begin_iter, begin_iter + verts.size());
+
+      verts.resize(thrust::distance(begin_iter, new_end), handle.get_stream());
+      labels->resize(thrust::distance(begin_iter, new_end), handle.get_stream());
+    } else {
+      thrust::sort(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      auto new_end = thrust::unique(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      verts.resize(thrust::distance(verts.begin(), new_end), handle.get_stream());
+    }
+  }
```

Comment:
Isn't this `thrust::set_union`? (https://nvidia.github.io/thrust/api/groups/group__set__operations.html)

Reply:
And the current code has a side effect of de-duplication (regardless of `dedupe_sources`). Is this intended?

Reply:
And if the intention is to remove the vertices already in `vertex_used_as_source`, shouldn't this be `set_difference` than `set_union`? We really need to make the expected behavior clear for all the possible combination of flags.

Reply:
This is processing the set of used vertices... we're not touching the frontier vertices yet.  So the de-duplication will be against the set of used vertices.

I will look at maintaining this list as sorted and doing a `set_union`.  The current construction of the `vertex_used_as_source` tuple is done just by appending, this step is to do the sort/unique that would be equivalent to `set_union`.